lib/os/sem.c

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <misc/sem.h>
+#include <syscall_handler.h>
+
+#ifdef CONFIG_USERSPACE
+void sys_sem_init(struct sys_sem *sem, unsigned int initial_count,
+		unsigned int limit)
+{
+	__ASSERT(limit != 0U, "limit cannot be zero");
+	__ASSERT(initial_count <= limit,
+			"count cannot be greater than limit");
+	__ASSERT(limit <= INT_MAX, "limit can't greater that INT_MAX");
+
+	atomic_set(&sem->futex.val, initial_count);
+	sem->limit = limit;
+}
+
+unsigned int sys_sem_count_get(struct sys_sem *sem)
+{
+	int value = atomic_get(&sem->futex.val);
+
+	return value > 0 ? value : 0;
+}
+
+int sys_sem_give(struct sys_sem *sem)
+{
+	int ret = 0;
+	int old_value;
+
+	do {
+		old_value = atomic_get(&sem->futex.val);
+		if (old_value >= sem->limit) {
+			break;
+		}
+	} while (atomic_cas(&sem->futex.val, old_value,
+			old_value < 0 ? 1 : old_value + 1) == 0);
+
+	if (old_value < 0) {
+		ret = k_futex_wake(&sem->futex, true);
+
+		if (ret > 0) {
+			return 0;
+		}
+	} else if (old_value >= sem->limit) {
+		return -EAGAIN;
+	}
+
+	return ret;
+}
+
+int sys_sem_take(struct sys_sem *sem, s32_t timeout)
+{
+	int ret = 0;
+	int old_value;
+
+	do {
+		do {
+			old_value = atomic_get(&sem->futex.val);
+			if (old_value < 0) {
+				break;
+			}
+		} while (atomic_cas(&sem->futex.val,
+				old_value, old_value - 1) == 0);
+
+		if (old_value > 0) {
+			return 0;
+		}
+
+		ret = k_futex_wait(&sem->futex, -1, timeout);
+	} while (ret == 0 || ret == -EAGAIN);
+
+	return ret;
+}
+#else
+void sys_sem_init(struct sys_sem *sem, unsigned int initial_count,
```

Comment:
There's no code coverage for these non-userspace versions.
Please add another scenario to test them. You can see what I did in tests/kernel/mutex/sys_mutex